app/validators/placeholders/placeholder_validator.py

```diff
@@ -108,7 +108,61 @@ def validate_option_label_from_value_placeholder(self, answer_id):
                 identifier=answer_id,
             )
 
-    def validate_placeholder_transforms(self, transforms):
+    def validate_answer_type_for_transform(
+        self, argument, argument_name, transform_type
+    ):
+        if not (
+            transform_type in ["format_unit", "format_percentage"]
+            and argument_name == "value"
+            and argument.get("value", {}).get("source") == "answers"
+        ):
+            return None
+
+        answer_id = argument["value"]["identifier"]
+        answer_type = self.questionnaire_schema.answers_with_context[answer_id][
+            "answer"
+        ]["type"]
+
+        if answer_type.lower() in transform_type:
+            return None
+
+        expected_type = transform_type.split("_")[1].title()
+        self.add_error(
+            error_messages.ANSWER_TYPE_FOR_TRANSFORM_TYPE_INVALID.format(
+                transform=transform_type,
+                expected_type=expected_type,
+                answer_type=answer_type,
+            ),
+            identifier=answer_id,
+        )
+
+    def validate_answer_and_transform_unit_match(self, *, arguments, transform_type):
+        if transform_type != "format_unit":
+            return None
+
+        value = arguments["value"]
+        answer_id = value.get("identifier")
+        unit = arguments["unit"]
+
+        if not self.errors and (
+            unit
+            == self.questionnaire_schema.answers_with_context[answer_id]["answer"][
+                "unit"
+            ]
+        ):
```

Comment:
This should be `if error or unit_match: return None`. This would check it if there was error. We want to exit early if there are errors or if there isnt error but unit does match.

Reply:
Added thanks!